keps/sig-multicluster/1645-multi-cluster-services-api/README.md

```diff
@@ -0,0 +1,925 @@
+
+```
+<<[UNRESOLVED re:stateful sets]>>
+  Today's headless services likely don't want a VIP and may not function
+  properly behind one. It probably doesn't make sense to export a current
+  headless service to the supercluster, it would work, but likely not the way
+  you want.
+<<[/UNRESOLVED]>>
+```
+- `NodePort` and `LoadBalancer`: These create `ClusterIP` services that would
+  sync as expected. For example If you export a `NodePort` service, the
+  resulting cross-cluster service will still be a supercluster IP type. You
+  could use node ports to access the cluster-local service in the source
+  cluster, but not in any other cluster, and it would only route to local
+  endpoints.
+- `ExternalName`: It doesn't make sense to export an `ExternalName` service.
+  They can't be merged with other exports, and it seems like it would only
+  complicate deployments by even attempting to stretch them across clusters.
+  Instead, regular `ExternalName` type `Services` should be created in each
+  cluster individually.
+
+### Consumption of EndpointSlice
+
+To consume a supercluster service, users will use the domain name associated
+with their `ServiceExport`. When the mcsd-controller sees a `ServiceExport`, a
+`ServiceImport` will be introduced, which can be largely ignored by the user. 
+
+An `ServiceImport` is a service that may have endpoints in other clusters.
+This includes 3 scenarios:
+1. This service is running entirely in different cluster(s)
+1. This service has endpoints in other cluster(s) and in this cluster
+1. This service is running entirely in this cluster, but is exported to other cluster(s) as well
+
+For each exported service, one `ServiceExport` will exist in each cluster that
+runs the service. The mcsd-controller will create and maintain a derived
+`ServiceImport` in each cluster within the supercluster (see: [constraints and
+conflict resolution](#constraints-and-conflict-resolution)). If all `ServiceExport` instances are deleted, each
+`ServiceImport` will also be deleted from all clusters.
+
+Since a given `ServiceImport` may be backed by multiple `EndpointSlices`, a
+given `EndpointSlice` will reference its `ServiceImport` using the label
+`multicluster.kubernetes.io/imported-service-name` similarly to how an
+`EndpointSlice` is associated with its `Service` in a single cluster. Each
+imported `EndpointSlice` will also have a
+`multicluster.kubernetes.io/source-cluster` label with a registry-scoped unique
+identifier for the cluster.
+
+```golang
+// ServiceImport declares that the specified service should be exported to other clusters.
+type ServiceImport struct {
+ metav1.TypeMeta `json:",inline"`
+ // +optional
+ metav1.ObjectMeta `json:"metadata,omitempty"`
+ // +optional
+ Spec ServiceImportSpec `json:"spec,omitempty"`
+}
+
+// ServiceImportSpec contains the current status of an imported service and the
+// information necessary to consume it
+type ServiceImportSpec struct {
```

Comment:
IF we support headless, you probably need some `type` here with options "VIP" or "SuperclusterIP" and "Headless" or "Direct" or something.

We also need rules that say what happens when cluster A exposes foo as a ClusterIP and cluster B exposes it as headless.  Probably ClusterIP wins and the headless cluster gets a status on their ServiceExport ?

Overall I think we need more status :)

Reply:
One way Headless can be supported is if instead of relying on `SuperclusterIP` we return a list of IPs for that service in each cluster and use DNS to return IP[s] from one of the clusters. In case of ClusterIP services it would be single IP in each cluster, in case of Headless it would be list of IPs. So we first check if SuperclusterIP is present in ServiceImport. If it is, use that. If not, look for and return IP[s] from one of the clusters.

Reply:
I think headless absolutely makes sense. I'll add a "Type" here so we can differentiate.

@vthapar I'm curious about this use case. This scenario seems somewhere between headless - where you know you're getting all backends, and VIP, where you know you're getting a consistent VIP. The benefits of a VIP are mainly in not having to rely on proper DNS implementation (which is historically not something we can expect) and frequent lookups. If we returned all `ClusterIP`s, this would mean that the client needs to be aware of this behavior. Some apps only do lookups once and don't respect TTLs, what would happen if the cluster they routed to went away or stopped exporting? Is there a reason you'd want to address a cluster specific Service, or would this be transparent to the consumer?

Reply:
We have an RFE on lighthouse for such a use case https://github.com/submariner-io/lighthouse/issues/117 I've asked them to provide more details on the use case. But from our brief interaction on slack the use case is they have headless service running in single cluster, clients need to connect to one of the pods directly and are aware that they'll get multiple IPs. I think we can support headless services with a caveat that for now they can only exist in one cluster. It fits use case of service running in one cluster accessible from multiple clusters.

Reply:
> The benefits of a VIP are mainly in not having to rely on proper DNS implementation (which is historically not something we can expect) and frequent lookups. If we returned all `ClusterIP`s, this would mean that the client needs to be aware of this behavior.

Also, I'd say this is again an implementation and use case detail. VIP practically necessitates a central MCSD controller for universal IPAM. We have https://submariner-io.github.io/ project which has a CoreDNS plugin, https://github.com/submariner-io/lighthouse which takes care of the DNS implementation. It is much easier to add new features to CoreDNS as plugins than it is to kube-proxy. IMO designing a spec so that it completely rules out some of the use cases is not a wise thing in long term.

Reply:
>I think we can support headless services with a caveat that for now they can only exist in one cluster

There are a few things we need to figure out for headless services, mainly around how we handle conflicting pod hostnames. The single-cluster constraint definitely simplifies things and is a good starting point, but I hope we can come up with a story that works for multiple clusters as well. I've put together some ideas for a followup PR.

> VIP practically necessitates a central MCSD controller for universal IPAM

I think we can avoid that - we absolutely don't want to add that constraint. This proposal allows a different VIP to be used for each cluster. A decentralized implementation could for example allocate the VIP locally from a cluster-specific range. It would even be technically compliant with this proposal (though with other drawbacks) to return one of the existing `ClusterIP`s as the supercluster VIP.

The DNS implementation issues I mentioned above are client side - lighthouse is an excellent answer to the server side, which is needed with a VIP or without. Many workloads/frameworks/languages historically do not respect TTLs, and the VIP approach solves that problem. A VIP/kube-proxy implementation also means there is only one component that needs to monitor service updates for all workloads and is consistent with the cluster-local model. Maintaining that behavior with supercluster services increases the likelihood that services will be portable without additional updates.

Reply:
> I think we can avoid that - we absolutely don't want to add that constraint. This proposal allows a different VIP to be used for each cluster. A decentralized implementation could for example allocate the VIP locally from a cluster-specific range. It would even be technically compliant with this proposal (though with other drawbacks) to return one of the existing `ClusterIP`s as the supercluster VIP.

We have something similar in submariner/lighthouse but for a different use case. That is definitely a doable option.

> The DNS implementation issues I mentioned above are client side - lighthouse is an excellent answer to the server side, which is needed with a VIP or without. Many workloads/frameworks/languages historically do not respect TTLs, and the VIP approach solves that problem. A VIP/kube-proxy implementation also means there is only one component that needs to monitor service updates for all workloads and is consistent with the cluster-local model. Maintaining that behavior with supercluster services increases the likelihood that services will be portable without additional updates.

Agree that DNS is not very client side friendly, depending on how applications are written. My take is that it should be an implementation detail and API shouldn't outright disallow one approach over another. Requiring applications to use "supercluster.local" already requires updates to work in multicluster env, unless plan is to use `resolve.conf` search path for it.

Reply:
> My take is that it should be an implementation detail and API shouldn't outright disallow one approach over another.

I want to keep this as flexible as we can, and dictate as little implementation as possible. The single VIP is an implementation detail, but it spills over into behavior. If we have it, we can say that any workload that works with MCS implementation X will continue to work with MCS implementation Y - actually, we can hopefully say that any workload that works with cluster.local services will work with MCS services. If we leave the VIP as an implementation detail, then only workloads which have properly behaving DNS clients can work with MCS implementation Y, regardless of how they fit with X. In that case, users need to be more aware of implementation details and the API becomes less portable.

> Requiring applications to use "supercluster.local" already requires updates to work in multicluster env

I agree, but this is likely to be a config change vs the code change that would be required to implement proper DNS handling in a workload that doesn't have it